ledger/eval.go

```diff
@@ -1049,6 +1084,47 @@ func (eval *BlockEvaluator) endOfBlock() error {
 	return nil
 }
 
+// modifyOfflineAccounts after all transactions and rewards are processed, modify the accounts so that their status is offline
+func (eval *BlockEvaluator) modifyOfflineAccounts() error {
+
+	proto, ok := config.Consensus[eval.block.BlockHeader.CurrentProtocol]
+	if !ok {
+		return fmt.Errorf("unknown consensus version: %s", eval.block.BlockHeader.CurrentProtocol)
+	}
+
+	expectedMaxNumberOfExpiredAccounts := proto.MaxExpiredAccountsToProcess
+	lengthOfExpiredParticipationAccounts := len(eval.block.ParticipationUpdates.ExpiredParticipationAccounts)
+
+	// If the length of the array is strictly greater than our max then we have an error.
+	// This works when the expected number of accounts is zero (i.e. it is disabled) as well
+	if lengthOfExpiredParticipationAccounts > expectedMaxNumberOfExpiredAccounts {
+		return fmt.Errorf("length of expired accounts (%d) was greater than expected (%d)",
+			lengthOfExpiredParticipationAccounts, expectedMaxNumberOfExpiredAccounts)
+	}
+
+	for _, accountAddr := range eval.block.ParticipationUpdates.ExpiredParticipationAccounts {
+		acctData, err := eval.state.lookup(accountAddr)
+		if err != nil {
+			return fmt.Errorf("modifyOfflineAccounts wrong: %v was not found", accountAddr)
+		}
+
+		// Reset the appropriate account data
+		acctData.Status = basics.Offline
+		acctData.VoteFirstValid = basics.Round(0)
+		acctData.VoteLastValid = basics.Round(0)
+		acctData.VoteKeyDilution = 0
+		acctData.VoteID = crypto.OneTimeSignatureVerifier{}
+		acctData.SelectionID = crypto.VRFVerifier{}
```

Comment:
I think that it would be cleaner if you were to create a `ClearOnlineState` on `AccountData` and call it from here. In the future, we might have additional fields, and it would make it cleaner having all of them in a single place.